In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Import Data
path= r'C:\Users\Zinnia\Documents\Career Foundry Data Analytics Course\Achievement 4\A4 Project\Data'

In [3]:
#Importing Data
df_ords_clean=pd.read_csv(r'C:\Users\Zinnia\Documents\Career Foundry Data Analytics Course\Achievement 4\A4 Project\Data\Prepared Data\orders_final.csv', index_col=False)

In [4]:
#Importing Data
df_ords_prior=pd.read_csv(r'C:\Users\Zinnia\Documents\Career Foundry Data Analytics Course\Achievement 4\A4 Project\Data\Prepared Data\order_products__prior.csv', index_col=False)

In [5]:
#Checking shape of data
df_ords_clean.shape

(3421083, 10)

In [6]:
#Checking shape of data
df_ords_prior.shape

(32434489, 4)

In [7]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [8]:
df_ords_clean.head()

Unnamed: 0  Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0           0             0   2539329        1    prior             1   
1           1             1   2398795        1    prior             2   
2           2             2    473747        1    prior             3   
3           3             3   2254736        1    prior             4   
4           4             4    431534        1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  first_order  
0                   2                  8                     NaN         True  
1                   3                  7                    15.0        False  
2                   3                 12                    21.0        False  
3                   4                  7                    29.0        False  
4                   4                 15                    28.0        False

In [9]:
#Merging Data for old orders and current orders
df_merged_large = df_ords_clean.merge(df_ords_prior, on = 'order_id', how='left', indicator = True)

In [10]:
df_merged_large['_merge'].value_counts()

both          32434489
left_only       206209
right_only           0
Name: _merge, dtype: int64

In [11]:
df_merged_large.shape

(32640698, 14)

#Number of rows validated as correct (32,640,698. Still have extra unnamed columns

In [12]:
df_merged_large.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'order_id', 'user_id', 'eval_set',
       'order_number', 'orders_day_of_week', 'order_hour_of_day',
       'days_since_prior_order', 'first_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge'],
      dtype='object')

In [13]:
#Drop unnamed columns
df_merged = df_merged_large.drop(columns = ['Unnamed: 0', 'Unnamed: 0.1'])

In [14]:
df_merged.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', '_merge'],
      dtype='object')

In [15]:
df_merged.shape

(32640698, 12)

In [16]:
df_merged

order_id  user_id eval_set  order_number  orders_day_of_week  \
0          2539329        1    prior             1                   2   
1          2539329        1    prior             1                   2   
2          2539329        1    prior             1                   2   
3          2539329        1    prior             1                   2   
4          2539329        1    prior             1                   2   
...            ...      ...      ...           ...                 ...   
32640693   2977660   206209    prior            13                   1   
32640694   2977660   206209    prior            13                   1   
32640695   2977660   206209    prior            13                   1   
32640696   2977660   206209    prior            13                   1   
32640697    272231   206209    train            14                   6   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True       196.0   
1                         8                     NaN         True     14084.0   
2                         8                     NaN         True     12427.0   
3                         8                     NaN         True     26088.0   
4                         8                     NaN         True     26405.0   
...                     ...                     ...          ...         ...   
32640693                 12                     7.0        False     38730.0   
32640694                 12                     7.0        False     31477.0   
32640695                 12                     7.0        False      6567.0   
32640696                 12                     7.0        False     22920.0   
32640697                 14                    30.0        False         NaN   

          add_to_cart_order  reordered     _merge  
0                       1.0        0.0       both  
1                       2.0        0.0       both  
2                       3.0        0.0       both  
3                       4.0        0.0       both  
4                       5.0        0.0       both  
...                     ...        ...        ...  
32640693                6.0        0.0       both  
32640694                7.0        0.0       both  
32640695                8.0        0.0       both  
32640696                9.0        0.0       both  
32640697                NaN        NaN  left_only  

[32640698 rows x 12 columns]

In [17]:
#Import Data to merge
df_products=pd.read_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'), index_col=False)

In [18]:
df_products

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

In [19]:
#Check shape of dataframe to merge 
df_products.shape

(49672, 6)

In [20]:
#Drop unnamed columns
df_merged_2 = df_merged.drop(columns = ['_merge'])

In [21]:
#Merge the products and orders dataframes to create a new dataframe
#Note: I first tried it with the dataframes in the other order and that caused the merge to not merge correctly
df_orders_products_merged = df_merged_2.merge(df_products, how= 'left', on= ['product_id'], indicator= True)

In [22]:
df_orders_products_merged.shape

(32641268, 17)

In [23]:
df_merged_2

order_id  user_id eval_set  order_number  orders_day_of_week  \
0          2539329        1    prior             1                   2   
1          2539329        1    prior             1                   2   
2          2539329        1    prior             1                   2   
3          2539329        1    prior             1                   2   
4          2539329        1    prior             1                   2   
...            ...      ...      ...           ...                 ...   
32640693   2977660   206209    prior            13                   1   
32640694   2977660   206209    prior            13                   1   
32640695   2977660   206209    prior            13                   1   
32640696   2977660   206209    prior            13                   1   
32640697    272231   206209    train            14                   6   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True       196.0   
1                         8                     NaN         True     14084.0   
2                         8                     NaN         True     12427.0   
3                         8                     NaN         True     26088.0   
4                         8                     NaN         True     26405.0   
...                     ...                     ...          ...         ...   
32640693                 12                     7.0        False     38730.0   
32640694                 12                     7.0        False     31477.0   
32640695                 12                     7.0        False      6567.0   
32640696                 12                     7.0        False     22920.0   
32640697                 14                    30.0        False         NaN   

          add_to_cart_order  reordered  
0                       1.0        0.0  
1                       2.0        0.0  
2                       3.0        0.0  
3                       4.0        0.0  
4                       5.0        0.0  
...                     ...        ...  
32640693                6.0        0.0  
32640694                7.0        0.0  
32640695                8.0        0.0  
32640696                9.0        0.0  
32640697                NaN        NaN  

[32640698 rows x 11 columns]

In [24]:
df_merged_2.shape

(32640698, 11)

In [25]:
df_products

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

In [26]:
df_products.product_id

0            1
1            2
2            3
3            4
4            5
         ...  
49667    49684
49668    49685
49669    49686
49670    49687
49671    49688
Name: product_id, Length: 49672, dtype: int64

In [27]:
df_products.product_id.nunique()

49670

In [28]:
df_products.product_id.isnull().sum()

0

In [29]:
df_merged_2.product_id.isnull().sum()

206209

In [30]:
df_merged_2.shape

(32640698, 11)

In [31]:
df_merged_2.shape[0]

32640698

In [32]:
df_orders_products_merged.shape[0]

32641268

In [33]:
df_orders_products_merged.shape[0]-df_merged_2.shape[0]

570

In [34]:
df_products.drop_duplicates(subset='product_id')

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49670 rows x 6 columns]

In [35]:
df_products=df_products.drop_duplicates(subset='product_id')

In [36]:
df_orders_products_merged = df_merged_2.merge(df_products, how= 'left', on= ['product_id'], indicator= True)

In [37]:
df_orders_products_merged.shape[0]

32640698

In [38]:
df_merged_2.shape[0]

32640698

In [39]:
help(df_products.drop_duplicates)

Help on method drop_duplicates in module pandas.core.frame:

drop_duplicates(subset: 'Hashable | Sequence[Hashable] | None' = None, keep: "Literal['first'] | Literal['last'] | Literal[False]" = 'first', inplace: 'bool' = False, ignore_index: 'bool' = False) -> 'DataFrame | None' method of pandas.core.frame.DataFrame instance
    Return DataFrame with duplicate rows removed.
    
    Considering certain columns is optional. Indexes, including time indexes
    are ignored.
    
    Parameters
    ----------
    subset : column label or sequence of labels, optional
        Only consider certain columns for identifying duplicates, by
        default use all of the columns.
    keep : {'first', 'last', False}, default 'first'
        Determines which duplicates (if any) to keep.
        - ``first`` : Drop duplicates except for the first occurrence.
        - ``last`` : Drop duplicates except for the last occurrence.
        - False : Drop all duplicates.
    inplace : bool, default False
  

In [40]:
df_orders_products_merged['_merge'].value_counts()

both          32404289
left_only       236409
right_only           0
Name: _merge, dtype: int64

#The merge was successful and has the correct number of rows (both= 32,404,859 and left_only=236,409)

In [41]:
#Export data in pickle format
df_orders_products_merged.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_may.pkl'))

In [43]:
df_orders_products_merged.shape

(32640698, 17)